In [2]:
import pandas as pd
import numpy as np
import sklearn

In [3]:
import pandas as pd

data = {
    'Date': ['2023-01-01', '2023-01-01', '2023-01-02', '2023-01-02', '2023-02-01', '2023-02-01', '2023-02-02', '2023-02-02', '2023-03-01', '2023-03-01', '2023-03-02', '2023-03-02'],
    'Product': ['A', 'B', 'B', 'A', 'A', 'B', 'A', 'B', 'B', 'A', 'A', 'B'],
    'Revenue': [221, 111, 171, 141, 75, 90, 96, 170, 200, 120, 121, 150],
    'Quantity': [3, 8, 11, 15, 4, 12, 17, 10, 19, 10, 12, 5]
}
df = pd.DataFrame(data)

print(df)


          Date Product  Revenue  Quantity
0   2023-01-01       A      221         3
1   2023-01-01       B      111         8
2   2023-01-02       B      171        11
3   2023-01-02       A      141        15
4   2023-02-01       A       75         4
5   2023-02-01       B       90        12
6   2023-02-02       A       96        17
7   2023-02-02       B      170        10
8   2023-03-01       B      200        19
9   2023-03-01       A      120        10
10  2023-03-02       A      121        12
11  2023-03-02       B      150         5


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Date      12 non-null     object
 1   Product   12 non-null     object
 2   Revenue   12 non-null     int64 
 3   Quantity  12 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 516.0+ bytes


In [7]:
df.nunique()

Date         6
Product      2
Revenue     12
Quantity    10
dtype: int64

## 1 Arrange DataFrame df, focusing on distinct product categories. Explore the data to reveal:

Overall revenue insights

In [14]:
p= df.groupby('Product')['Revenue'].sum().reset_index()
q= p.sort_values(by='Revenue', ascending=True)
print(q)

  Product  Revenue
0       A      774
1       B      892


In [15]:
p= df.groupby('Product')['Revenue'].sum().reset_index()
q= p.sort_values(by='Revenue', ascending=False)
print(q)

  Product  Revenue
1       B      892
0       A      774


the overall revenue insights is 774 for product 'A' and 892 for product 'B',by this we can observe that the revenue of product B is greater than product A in the given time period

Average revenue patterns

In [ ]:
df.groupby(['Product'])['Revenue'].mean()

Peak quantity sold observations

In [17]:
df.groupby(['Product'])['Revenue'].max()

Product
A    221
B    200
Name: Revenue, dtype: int64

Minimum quantity sold occurrences

In [21]:
k =s=df.groupby(['Product'])['Quantity'].min()

k

Product
A    3
B    5
Name: Quantity, dtype: int64

2) Aggregate DataFrame df by extracting the month and year from the 'Date' column. Calculate the collective revenue for each unique combination of month and year.

The pd.to_datetime function in pandas is used to convert argument to datetime. In the context of a DataFrame, it's often used to convert a column containing date-like strings or other date-like objects into a pandas datetime object, which allows for convenient datetime operations.

In [23]:
df['Date'] = pd.to_datetime(df['Date'])
df['Month'] = df['Date'].dt.month
df['Year'] = df['Date'].dt.year
df[['Date','Month','Year']]

Date  Month  Year
0  2023-01-01      1  2023
1  2023-01-01      1  2023
2  2023-01-02      1  2023
3  2023-01-02      1  2023
4  2023-02-01      2  2023
5  2023-02-01      2  2023
6  2023-02-02      2  2023
7  2023-02-02      2  2023
8  2023-03-01      3  2023
9  2023-03-01      3  2023
10 2023-03-02      3  2023
11 2023-03-02      3  2023

In [25]:
x = df.groupby(['Month', 'Year'])['Revenue'].sum()
print(x)

Month  Year
1      2023    644
2      2023    431
3      2023    591
Name: Revenue, dtype: int64


In [27]:
df['Date'] = pd.to_datetime(df['Date'])
df['MonthYear'] = df['Date'].dt.to_period("M")
revenue_by_month_year = df.groupby('MonthYear')['Revenue'].sum().reset_index()
print(revenue_by_month_year)

  MonthYear  Revenue
0   2023-01      644
1   2023-02      431
2   2023-03      591


3)Group the DataFrame df by the 'Date' column and find the date with the highest total revenue and the date with the highest total quantity sold.

In [29]:
h= df.groupby('Date').agg({'Revenue': 'sum', 'Quantity': 'sum'}).reset_index()
h

Date  Revenue  Quantity
0 2023-01-01      332        11
1 2023-01-02      312        26
2 2023-02-01      165        16
3 2023-02-02      266        27
4 2023-03-01      320        29
5 2023-03-02      271        17

In [40]:
grouped_by_date = h.groupby('Date')['Revenue'].sum()
grouped_by_date


Date
2023-01-01    332
2023-01-02    312
2023-02-01    165
2023-02-02    266
2023-03-01    320
2023-03-02    271
Name: Revenue, dtype: int64

In [41]:
max_revenue_date = grouped_by_date.idxmax()
print(f"Date with highest total revenue: {max_revenue_date}")

Date with highest total revenue: 2023-01-01 00:00:00


idxmax():
In pandas, the idxmax() function is used to get the index of the first occurrence of the maximum value in a Series.
idxmax() is used to find the index of the maximum value in the 'Revenue' and 'Quantity' columns within the grouped DataFrame (hr). Then, loc is used to locate the corresponding 'Date' value for that index.

4)Group the Data Frame df by the Date and find products that have a total quantity sold greater than 10 and less than 15 .

In [43]:
x = df.groupby(['Date', 'Product'])['Quantity'].sum()

In [44]:
y = x[(x > 10) & (x < 15)]


In [45]:
x,y

(Date        Product
 2023-01-01  A           3
             B           8
 2023-01-02  A          15
             B          11
 2023-02-01  A           4
             B          12
 2023-02-02  A          17
             B          10
 2023-03-01  A          10
             B          19
 2023-03-02  A          12
             B           5
 Name: Quantity, dtype: int64,
 Date        Product
 2023-01-02  B          11
 2023-02-01  B          12
 2023-03-02  A          12
 Name: Quantity, dtype: int64)

5)Group the DataFrame df by the 'Product' column and add a new column 'Revenue Percentage' that shows the percentage of each product's revenue relative to the total revenue for that product.

In [46]:
x = df.groupby('Product')['Revenue'].sum()

In [47]:
x

Product
A    774
B    892
Name: Revenue, dtype: int64

In [48]:
percentage = x / x.sum() * 100

In [ ]:
df['Revenue Percentage'] = gbt['Product'].map(percentage)
print()